From: Stephen Gran <sgran@debian.org>
Date: Wed, 27 Oct 2010 09:07:26 +0200
Subject: [PATCH] dovecot-ssl

Description: provide mechanism to discover if ssl client certificate is verified
Bug-Debian: http://bugs.debian.org/446555
Origin: vendor, http://bugs.debian.org/cgi-bin/bugreport.cgi?msg=25;filename=dovecot-ssl.patch;att=1;bug=446555
---
 src/auth/auth-request.c              |    2 ++
 src/auth/auth-request.h              |    1 +
 src/auth/db-checkpassword.c          |    4 ++++
 src/lib-auth/auth-client.h           |    1 +
 src/lib-auth/auth-server-request.c   |   12 +++++++++++-
 src/login-common/sasl-server.c       |    2 ++
 src/login-common/ssl-proxy-gnutls.c  |    5 +++++
 src/login-common/ssl-proxy-openssl.c |   31 +++++++++++++++++++++++++++++++
 src/login-common/ssl-proxy.c         |    6 ++++++
 src/login-common/ssl-proxy.h         |    1 +
 10 files changed, 64 insertions(+), 1 deletions(-)

diff --git a/src/auth/auth-request.c b/src/auth/auth-request.c
index 96ea0da..2380201 100644
--- a/src/auth/auth-request.c
+++ b/src/auth/auth-request.c
@@ -166,6 +166,8 @@ bool auth_request_import(struct auth_request *request,
 		request->master_user = p_strdup(request->pool, value);
 	else if (strcmp(key, "original_username") == 0)
 		request->original_username = p_strdup(request->pool, value);
+	else if (strcmp(key, "cert_fingerprint") == 0)
+		request->cert_fingerprint = p_strdup(request->pool, value);
 	else if (strcmp(key, "cert_username") == 0) {
 		if (request->auth->ssl_username_from_cert) {
 			/* get username from SSL certificate. it overrides
diff --git a/src/auth/auth-request.h b/src/auth/auth-request.h
index b02ff15..e6adac9 100644
--- a/src/auth/auth-request.h
+++ b/src/auth/auth-request.h
@@ -44,6 +44,7 @@ struct auth_request {
 	const char *realm;
 	char *mech_password; /* set if verify_plain() is called */
 	char *passdb_password; /* set after password lookup if successful */
+	char *cert_fingerprint;
         /* extra_fields are returned in authentication reply. Fields prefixed
            with "userdb_" are skipped. If prefetch userdb is used, it uses
            the "userdb_" prefixed fields. */
diff --git a/src/auth/db-checkpassword.c b/src/auth/db-checkpassword.c
index 95191c6..146f1a7 100644
--- a/src/auth/db-checkpassword.c
+++ b/src/auth/db-checkpassword.c
@@ -120,6 +120,10 @@ void checkpassword_setup_env(struct auth_request *request)
 		env_put(t_strconcat("MASTER_USER=",
 				    request->master_user, NULL));
 	}
+	if (request->cert_fingerprint != NULL) {
+		env_put(t_strconcat("SSL_FINGERPRINT=",
+				request->cert_fingerprint, NULL));
+	}
 	if (request->extra_fields != NULL) {
 		const char *fields =
 			auth_stream_reply_export(request->extra_fields);
diff --git a/src/lib-auth/auth-client.h b/src/lib-auth/auth-client.h
index 267bb45..efab89e 100644
--- a/src/lib-auth/auth-client.h
+++ b/src/lib-auth/auth-client.h
@@ -26,6 +26,7 @@ struct auth_request_info {
 	const char *mech;
 	const char *service;
 	const char *cert_username;
+	const char *cert_fingerprint;
 	enum auth_request_flags flags;
 
 	struct ip_addr local_ip, remote_ip;
diff --git a/src/lib-auth/auth-server-request.c b/src/lib-auth/auth-server-request.c
index 2494b0c..ee29bc4 100644
--- a/src/lib-auth/auth-server-request.c
+++ b/src/lib-auth/auth-server-request.c
@@ -15,7 +15,7 @@ struct auth_request {
 
 	unsigned int id;
 
-	char *mech, *service, *cert_username;
+	char *mech, *service, *cert_username, *cert_fingerprint;
         enum auth_request_flags flags;
 	struct ip_addr local_ip, remote_ip;
 	unsigned int local_port, remote_port;
@@ -106,6 +106,14 @@ static int auth_server_send_new_request(struct auth_server_connection *conn,
 		}
 		str_printfa(str, "\tcert_username=%s", request->cert_username);
 	}
+	if (request->cert_fingerprint != NULL) {
+		if (!is_valid_string(request->cert_fingerprint)) {
+			t_pop();
+			*error_r = "Invalid fingerprint in SSL certificate";
+			return -1;
+		}
+		str_printfa(str, "\tcert_fingerprint=%s", request->cert_fingerprint);
+	}
 	if (request->local_ip.family != 0)
 		str_printfa(str, "\tlip=%s", net_ip2addr(&request->local_ip));
 	if (request->remote_ip.family != 0)
@@ -348,6 +356,7 @@ auth_client_request_new(struct auth_client *client, struct auth_connect_id *id,
 	request->mech = i_strdup(request_info->mech);
 	request->service = i_strdup(request_info->service);
 	request->cert_username = i_strdup(request_info->cert_username);
+	request->cert_fingerprint = i_strdup(request_info->cert_fingerprint);
 	request->flags = request_info->flags;
 	request->local_ip = request_info->local_ip;
 	request->remote_ip = request_info->remote_ip;
@@ -393,6 +402,7 @@ static void auth_client_request_free(struct auth_request *request)
 	i_free(request->mech);
 	i_free(request->service);
 	i_free(request->cert_username);
+	i_free(request->cert_fingerprint);
 	i_free(request);
 }
 
diff --git a/src/login-common/sasl-server.c b/src/login-common/sasl-server.c
index 9b9751d..152d0b6 100644
--- a/src/login-common/sasl-server.c
+++ b/src/login-common/sasl-server.c
@@ -163,6 +163,8 @@ void sasl_server_auth_begin(struct client *client,
 	info.service = service;
 	info.cert_username = client->proxy == NULL ? NULL :
 		ssl_proxy_get_peer_name(client->proxy);
+	info.cert_fingerprint = client->proxy == NULL ? NULL :
+		ssl_proxy_get_fingerprint(client->proxy);
 	info.flags = client_get_auth_flags(client);
 	info.local_ip = client->local_ip;
 	info.remote_ip = client->ip;
diff --git a/src/login-common/ssl-proxy-gnutls.c b/src/login-common/ssl-proxy-gnutls.c
index 402d712..2d6aa74 100644
--- a/src/login-common/ssl-proxy-gnutls.c
+++ b/src/login-common/ssl-proxy-gnutls.c
@@ -61,6 +61,11 @@ static const char *get_alert_text(struct ssl_proxy *proxy)
 	return gnutls_alert_get_name(gnutls_alert_get(proxy->session));
 }
 
+const char *ssl_proxy_get_fingerprint(struct ssl_proxy *proxy __attr_unused__)
+{
+	return NULL;
+}
+
 static int handle_ssl_error(struct ssl_proxy *proxy, int error)
 {
 	if (!gnutls_error_is_fatal(error)) {
diff --git a/src/login-common/ssl-proxy-openssl.c b/src/login-common/ssl-proxy-openssl.c
index 72fa0fa..b00a4b8 100644
--- a/src/login-common/ssl-proxy-openssl.c
+++ b/src/login-common/ssl-proxy-openssl.c
@@ -26,6 +26,8 @@
 /* Check every 30 minutes if parameters file has been updated */
 #define SSL_PARAMFILE_CHECK_INTERVAL (60*30)
 
+static const char hexcodes[] = "0123456789ABCDEF";
+
 enum ssl_io_action {
 	SSL_ADD_INPUT,
 	SSL_REMOVE_INPUT,
@@ -622,6 +624,35 @@ const char *ssl_proxy_get_peer_name(struct ssl_proxy *proxy)
 	return *name == '\0' ? NULL : name;
 }
 
+const char *ssl_proxy_get_fingerprint(struct ssl_proxy *proxy) {
+	X509 *x509;
+	char *peer_fingerprint = NULL;
+	unsigned char md[EVP_MAX_MD_SIZE];
+	unsigned int n;
+	int j;
+
+	if (!ssl_proxy_has_valid_client_cert(proxy))
+	return NULL;
+
+	x509 = SSL_get_peer_certificate(proxy->ssl);
+	if (x509 == NULL)
+		return NULL; /* we should have had it.. */
+
+	if (X509_digest(x509, EVP_md5(), md, &n) && n > 0) {
+		peer_fingerprint = i_malloc(n * 3);
+		for (j = 0; j < (int) n; j++) {
+			peer_fingerprint[j * 3] = hexcodes[(md[j] & 0xf0) >> 4U];
+			peer_fingerprint[(j * 3) + 1] = hexcodes[(md[j] & 0x0f)];
+			if (j + 1 != (int) n)
+				peer_fingerprint[(j * 3) + 2] = ':';
+			else
+				peer_fingerprint[(j * 3) + 2] = '\0';
+		}
+		i_info("fingerprint: %s", peer_fingerprint);
+	}
+	return (const char *)peer_fingerprint;
+}
+
 bool ssl_proxy_is_handshaked(const struct ssl_proxy *proxy)
 {
 	return proxy->handshaked;
diff --git a/src/login-common/ssl-proxy.c b/src/login-common/ssl-proxy.c
index 6a70b90..e7668f1 100644
--- a/src/login-common/ssl-proxy.c
+++ b/src/login-common/ssl-proxy.c
@@ -57,6 +57,12 @@ const char *ssl_proxy_get_security_string(struct ssl_proxy *proxy ATTR_UNUSED)
 
 void ssl_proxy_free(struct ssl_proxy *proxy ATTR_UNUSED) {}
 
+const char *ssl_proxy_get_fingerprint(struct ssl_proxy *proxy __attr_unused__)
+{
+	return NULL;
+}
+
+
 unsigned int ssl_proxy_get_count(void)
 {
 	return 0;
diff --git a/src/login-common/ssl-proxy.h b/src/login-common/ssl-proxy.h
index 11e3478..a1fdd14 100644
--- a/src/login-common/ssl-proxy.h
+++ b/src/login-common/ssl-proxy.h
@@ -25,6 +25,7 @@ bool ssl_proxy_is_handshaked(const struct ssl_proxy *proxy) ATTR_PURE;
 const char *ssl_proxy_get_last_error(const struct ssl_proxy *proxy) ATTR_PURE;
 const char *ssl_proxy_get_security_string(struct ssl_proxy *proxy);
 void ssl_proxy_free(struct ssl_proxy *proxy);
+const char *ssl_proxy_get_fingerprint(struct ssl_proxy *proxy);
 
 /* Return number of active SSL proxies */
 unsigned int ssl_proxy_get_count(void) ATTR_PURE;
-- 
